File: 007.跨端APP应用（移动端开发工程师）/uniapp-project/utils/category-presentation.ts
/**
 * 后端 category → 前端展示层派生。
 *
 * 所有分类的颜色 + 图标全部走后端存的 `c.color` / `c.icon`,前端不再维护预设表。
 * 这样所有端(uniapp / Harmony / Flutter)都用同一份后端数据,保证一处改色,各端跟随。
 *
 * 兜底:后端字段为空 / 无 category → FALLBACK_COLOR + FALLBACK_ICON。
 */
import type { Category } from '@/api/categories'

export interface CategoryPresentation {
  /** emoji 图标(后端存) */
  icon: string
  /** 分类 hex 颜色(后端存) */
  color: string
}

// 跨端安全占位
export const FALLBACK_ICON = '⋯'
export const FALLBACK_COLOR = '#727782'

// 检测 Material Symbols 风格的 ligature 名(如 'restaurant' / 'fastfood' / 'more_horiz')。
// 微信小程序/H5 默认不加载 MS Outlined 字体,直接渲染会显示字面文字,当作无效输入处理。
function isProbablyMsLigatureName(s: string): boolean {
  return /^[a-zA-Z0-9_]+$/.test(s)
}

export function categoryPresentation(c: Pick<Category, 'id' | 'type' | 'name'>): CategoryPresentation {
  const icon = (c as Partial<Category>).icon
  const color = (c as Partial<Category>).color
  const safeIcon = icon && !isProbablyMsLigatureName(icon) ? icon : FALLBACK_ICON
  const safeColor = color && color.length > 0 ? color : FALLBACK_COLOR
  return { icon: safeIcon, color: safeColor }
}

/** 仅取图标名(给已经自己用 cat.color 的场景,例如 breakdown 卡) */
export function categoryMaterialIcon(c: Pick<Category, 'id' | 'type' | 'name'>): string {
  return categoryPresentation(c).icon
}

File: 007.跨端APP应用（移动端开发工程师）/flutter-project/lib/core/utils/category_presentation.dart
import '../api/models.dart';

/// 对齐 utils/category-presentation.ts — 分类显示辅助。
///
/// 所有分类的颜色 + 图标全部走后端存的 `c.color` / `c.icon`,前端不再维护预设表。
/// 这样所有端(uniapp / Harmony / Flutter)都用同一份后端数据,
/// 保证一处改色,各端跟随。
class CategoryPresentation {
  const CategoryPresentation({required this.icon, required this.color});
  final String icon;
  final String color;
}

const _FALLBACK_ICON = '⋯';
const _FALLBACK_COLOR = '#727782';

bool _isProbablyMsLigatureName(String s) {
  // 老分类用 MS 字体 ligature(如 'home')写 icon,跨端表现不一,fallback 到 emoji。
  return RegExp(r'^[a-zA-Z0-9_\-]+$').hasMatch(s);
}

CategoryPresentation presentCategory(Category c) {
  var icon = c.icon.trim();
  if (icon.isEmpty || _isProbablyMsLigatureName(icon)) {
    icon = _FALLBACK_ICON;
  }
  final color = c.color.isEmpty ? _FALLBACK_COLOR : c.color;
  return CategoryPresentation(icon: icon, color: color);
}

String categoryMaterialIcon(Category c) => presentCategory(c).icon;
